File: packages/canvas-react/src/io/tableMerge.ts
/**
 * Converging a table's two representations: `rows` and `sheet`.
 *
 * A table artifact carries the same data twice once a person has edited it:
 * the structured `columns`/`rows` (what agents read and write) and the
 * Fortune-sheet editor state `data.sheet` (what the grid renders — values
 * plus formatting, merges and typed formulas). Both accept writes, so the
 * two drift apart unless something reconciles them. This module is that
 * something, in both directions:
 *
 * - {@link projectSheetIntoRows} (person → rows): runs at save time, so the
 *   stored `rows` always reflect what the person sees. This establishes the
 *   invariant "at every save, rows equal the sheet's data rectangle" — the
 *   common ancestor that makes the merge below conflict-free.
 * - {@link mergeRowsIntoSheet} (agent → sheet): runs wherever the sheet is
 *   consumed (render, export). Starting from the sheet, any rectangle cell
 *   whose rows value differs is overridden with the rows value — under the
 *   invariant, "differs" means "the agent wrote it after the last save", so
 *   rows win. Style keys (bold, colors, merges, …) are preserved.
 *
 * Known limits, by design:
 * - Cells outside the rows rectangle (notes a person typed beside the
 *   table) are preserved in the sheet but stay invisible to `rows`, CSV
 *   and agents.
 * - Agent row deletion is propagated by truncating trailing sheet rows
 *   (their formatting goes with them); a mid-table row insert/delete keeps
 *   every value correct but leaves row-level formatting at its old index.
 * - Agent column deletion is not propagated (the sheet keeps the column).
 * - Only the first sheet is reconciled — `rows` model a single grid.
 */

import type { TableColumn, TableData } from "../protocol/artifacts";
import type { FormulaValues } from "./formula";

type SheetCell = { r: number; c: number; v?: unknown };
type Sheet = Record<string, unknown> & { celldata?: SheetCell[]; data?: unknown[][] };

const isFormula = (v: unknown): v is string => typeof v === "string" && v.startsWith("=");

/** Keys of a Fortune cell object that carry the value; the rest is styling. */
const VALUE_KEYS = ["v", "m", "ct", "f", "qp", "spl"] as const;

/** Normalize a cell value for comparison: empty forms fold together and a
 * numeric string equals its number, so `80` vs `"80"` never reads as a
 * change (a false "difference" would overwrite on every load). */
function normalizedValue(value: unknown): string | null {
  if (value === null || value === undefined || value === "") return null;
  if (typeof value === "number") return Number.isFinite(value) ? String(value) : null;
  if (typeof value === "boolean") return value ? "TRUE" : "FALSE";
  const text = String(value).trim();
  if (text === "") return null;
  const num = Number(text);
  return Number.isNaN(num) ? text : String(num);
}

const normalizedFormula = (f: unknown): string | null =>
  typeof f === "string" && f.trim() ? f.trim().replace(/^=/, "") : null;

/** True when a rows value and a sheet cell hold the same content. */
export function sameCellContent(rowsValue: unknown, cellV: unknown): boolean {
  const cellFormula = cellV && typeof cellV === "object" ? normalizedFormula((cellV as { f?: unknown }).f) : null;
  if (isFormula(rowsValue) || cellFormula !== null) {
    return normalizedFormula(rowsValue) === cellFormula;
  }
  const cellValue =
    cellV && typeof cellV === "object"
      ? ((cellV as { v?: unknown; m?: unknown }).v ?? (cellV as { m?: unknown }).m ?? null)
      : cellV;
  return normalizedValue(rowsValue) === normalizedValue(cellValue);
}

/** The sheet's cells as a map keyed `"r,c"` (accepts celldata or the dense
 * data-matrix form Fortune emits from onChange). */
function cellMap(sheet: Sheet): Map<string, unknown> {
  const map = new Map<string, unknown>();
  if (sheet.celldata?.length) {
    for (const cell of sheet.celldata) {
      if (cell?.v != null) map.set(`${cell.r},${cell.c}`, cell.v);
    }
    return map;
  }
  (sheet.data ?? []).forEach((row, r) =>
    (row ?? []).forEach((cell, c) => {
      if (cell != null) map.set(`${r},${c}`, cell);
    }),
  );
  return map;
}

/** A sheet cell's content, for projecting into rows: the typed formula when
 * present (`f` wins over the cached value — a formula must stay a formula in
 * rows, or agents read it as a constant), else the value. */
function projectedValue(cellV: unknown): string | number {
  if (cellV === null || cellV === undefined) return "";
  if (typeof cellV !== "object") return cellV as string | number;
  const f = normalizedFormula((cellV as { f?: unknown }).f);
  if (f !== null) return `=${f}`;
  const v = (cellV as { v?: unknown; m?: unknown }).v ?? (cellV as { m?: unknown }).m;
  return v === null || v === undefined ? "" : (v as string | number);
}

/**
 * Person → rows: rows rebuilt from the sheet's data rectangle (display row 1
 * onward, one column per `columns` entry). Row keys outside `columns` are
 * carried over from the existing rows untouched. Trailing rows with no
 * content in the rectangle are dropped (a person deleting rows shrinks the
 * table); the row count otherwise follows the sheet's deepest content row,
 * so person-added rows extend `rows`.
 */
export function projectSheetIntoRows(
  columns: TableColumn[],
  rows: TableData["rows"],
  sheet: TableData["sheet"],
): TableData["rows"] {
  const first = sheet?.[0] as Sheet | undefined;
  if (!first || columns.length === 0) return rows;
  const cells = cellMap(first);

  let deepest = 0;
  for (const key of cells.keys()) {
    const [r, c] = key.split(",").map(Number);
    if (r >= 1 && c < columns.length) deepest = Math.max(deepest, r);
  }

  const projected: TableData["rows"] = [];
  for (let i = 0; i < deepest; i++) {
    const row: Record<string, string | number> = { ...(rows[i] ?? {}) };
    columns.forEach((col, c) => {
      row[col.key] = projectedValue(cells.get(`${i + 1},${c}`));
    });
    projected.push(row);
  }
  return projected;
}

/** The Fortune value keys for one rows value (cached formula results come
 * from `formulas`, keyed like `computeFormulas` output). */
function valueCell(
  rowsValue: unknown,
  dataIdx: number,
  col: number,
  formulas: FormulaValues | undefined,
): Record<string, unknown> | null {
  if (rowsValue === null || rowsValue === undefined || rowsValue === "") return null;
  if (isFormula(rowsValue)) {
    const out: Record<string, unknown> = { f: rowsValue };
    const computed = formulas?.get(`${dataIdx + 1},${col}`);
    if (computed !== undefined) {
      out.v = computed;
      out.m = String(computed);
      if (typeof computed === "number") out.ct = { fa: "General", t: "n" };
    }
    return out;
  }
  const numeric = typeof rowsValue === "number";
  return { v: rowsValue, m: String(rowsValue), ...(numeric ? { ct: { fa: "General", t: "n" } } : {}) };
}

/**
 * Agent → sheet: the sheet with every rectangle cell whose rows value
 * differs overridden by the rows value. Style keys on the overridden cell
 * are preserved; only value keys change. Trailing sheet rows beyond
 * `rows.length` are dropped inside the rectangle (agent row deletion —
 * their formatting goes with them; out-of-rectangle cells stay). A no-op
 * (returns the input array) when nothing differs, so mounts stay stable.
 */
export function mergeRowsIntoSheet(
  columns: TableColumn[],
  rows: TableData["rows"],
  sheet: TableData["sheet"],
  formulas?: FormulaValues,
): TableData["sheet"] {
  const first = sheet?.[0] as Sheet | undefined;
  if (!first || columns.length === 0) return sheet;
  const cells = cellMap(first);

  const overrides = new Map<string, Record<string, unknown> | null>();
  columns.forEach((col, c) => {
    // Header row follows the column labels.
    const label = col.label ?? col.key;
    if (!sameCellContent(label, cells.get(`0,${c}`))) {
      overrides.set(`0,${c}`, { v: label, m: String(label) });
    }
    rows.forEach((row, i) => {
      const value = row[col.key];
      if (!sameCellContent(value, cells.get(`${i + 1},${c}`))) {
        overrides.set(`${i + 1},${c}`, valueCell(value, i, c, formulas));
      }
    });
  });

  const inRectangle = (r: number, c: number) => c < columns.length && r >= 1;
  const ghosts = [...cells.keys()].filter((key) => {
    const [r, c] = key.split(",").map(Number);
    return inRectangle(r, c) && r > rows.length;
  });
  if (overrides.size === 0 && ghosts.length === 0) return sheet;

  const merged: SheetCell[] = [];
  const seen = new Set<string>();
  for (const [key, existing] of cells) {
    const [r, c] = key.split(",").map(Number);
    seen.add(key);
    if (inRectangle(r, c) && r > rows.length) continue; // agent-deleted row
    if (!overrides.has(key)) {
      merged.push({ r, c, v: existing });
      continue;
    }
    const next = overrides.get(key);
    const style =
      existing && typeof existing === "object"
        ? Object.fromEntries(
            Object.entries(existing as Record<string, unknown>).filter(
              ([k]) => !(VALUE_KEYS as readonly string[]).includes(k),
            ),
          )
        : {};
    const v = next === null ? style : { ...style, ...next };
    if (Object.keys(v).length > 0) merged.push({ r, c, v });
  }
  for (const [key, next] of overrides) {
    if (seen.has(key) || next === null) continue;
    const [r, c] = key.split(",").map(Number);
    merged.push({ r, c, v: next });
  }
  merged.sort((a, b) => a.r - b.r || a.c - b.c);

  const { data: _data, ...rest } = first;
  const mergedFirst: Sheet = {
    ...rest,
    celldata: merged,
    row: Math.max(Number(first.row) || 0, rows.length + 40),
    column: Math.max(Number(first.column) || 0, columns.length + 2),
  };
  return [mergedFirst, ...(sheet ?? []).slice(1)] as TableData["sheet"];
}

File: packages/canvas-py/src/langchain_canvas/table_merge.py
"""Converging a table's two representations: ``rows`` and ``sheet``.

A table artifact carries the same data twice once a person has edited it:
the structured ``columns``/``rows`` (what agents read and write) and the
grid editor state ``data["sheet"]`` (what the person sees — values plus
formatting, merges and typed formulas). Both accept writes, so the two
drift apart unless something reconciles them. This module is the Python
side of that reconciliation, mirroring ``canvas-react/src/io/tableMerge.ts``:

- :func:`project_sheet_into_rows` (person → rows): runs at save time (the
  save endpoints call it through :func:`~langchain_canvas.replay.encode_artifact`),
  so stored ``rows`` always reflect what the person sees — including typed
  formulas, projected as their source text, never their cached value. This
  establishes the invariant "at every save, rows equal the sheet's data
  rectangle".
- :func:`merge_rows_into_sheet` (agent → sheet): runs where the sheet is
  consumed (the xlsx exporter). Under the invariant, a rectangle cell whose
  rows value differs was written by the agent after the last save, so the
  rows value wins; the cell's styling keys are preserved.

Known limits, by design: cells outside the rows rectangle (margin notes)
are preserved but stay invisible to ``rows``/CSV/agents; agent row deletion
truncates trailing sheet rows (their formatting goes too) while a mid-table
insert/delete keeps values right but leaves row-level formatting at its old
index; agent column deletion is not propagated; only the first sheet is
reconciled.
"""

from __future__ import annotations

from typing import Any

#: Keys of a grid cell object that carry the value; the rest is styling.
_VALUE_KEYS = ("v", "m", "ct", "f", "qp", "spl")


def _is_formula(value: Any) -> bool:
    return isinstance(value, str) and value.startswith("=")


def _normalized_value(value: Any) -> str | None:
    """Comparison form: empty forms fold together, numeric strings equal
    their numbers — ``80`` vs ``"80"`` must never read as a change."""
    if value is None or value == "":
        return None
    if isinstance(value, bool):
        return "TRUE" if value else "FALSE"
    if isinstance(value, (int, float)):
        return _number_text(float(value))
    text = str(value).strip()
    if not text:
        return None
    try:
        return _number_text(float(text))
    except ValueError:
        return text


def _number_text(num: float) -> str:
    return str(int(num)) if num == int(num) else str(num)


def _normalized_formula(f: Any) -> str | None:
    if isinstance(f, str) and f.strip():
        return f.strip().lstrip("=")
    return None


def same_cell_content(rows_value: Any, cell_v: Any) -> bool:
    """True when a rows value and a sheet cell hold the same content."""
    cell_formula = _normalized_formula(cell_v.get("f")) if isinstance(cell_v, dict) else None
    if _is_formula(rows_value) or cell_formula is not None:
        return _normalized_formula(rows_value) == cell_formula
    if isinstance(cell_v, dict):
        cell_value = cell_v.get("v") if cell_v.get("v") is not None else cell_v.get("m")
    else:
        cell_value = cell_v
    return _normalized_value(rows_value) == _normalized_value(cell_value)


def cell_map(sheet: dict[str, Any]) -> dict[tuple[int, int], Any]:
    """The sheet's cells keyed ``(r, c)`` — from ``celldata`` or the dense
    ``data`` matrix the grid emits from live state."""
    cells: dict[tuple[int, int], Any] = {}
    celldata = sheet.get("celldata")
    if celldata:
        for cell in celldata:
            if isinstance(cell, dict) and cell.get("v") is not None:
                cells[(int(cell["r"]), int(cell["c"]))] = cell["v"]
        return cells
    for r, row in enumerate(sheet.get("data") or []):
        for c, cell in enumerate(row or []):
            if cell is not None:
                cells[(r, c)] = cell
    return cells


def _projected_value(cell_v: Any) -> Any:
    """A sheet cell's content for rows: the typed formula when present
    (``f`` wins — a formula must stay a formula in rows), else the value."""
    if cell_v is None:
        return ""
    if not isinstance(cell_v, dict):
        return cell_v
    formula = _normalized_formula(cell_v.get("f"))
    if formula is not None:
        return f"={formula}"
    value = cell_v.get("v") if cell_v.get("v") is not None else cell_v.get("m")
    return "" if value is None else value


def project_sheet_into_rows(
    columns: list[dict[str, Any]],
    rows: list[dict[str, Any]],
    sheet: list[dict[str, Any]] | None,
) -> list[dict[str, Any]]:
    """Person → rows: rows rebuilt from the sheet's data rectangle.

    Row keys outside ``columns`` are carried over untouched; trailing rows
    with no rectangle content are dropped (person-deleted), person-added
    rows extend the list.
    """
    first = sheet[0] if sheet else None
    if not isinstance(first, dict) or not columns:
        return rows
    cells = cell_map(first)

    deepest = 0
    for r, c in cells:
        if r >= 1 and c < len(columns):
            deepest = max(deepest, r)

    projected: list[dict[str, Any]] = []
    for i in range(deepest):
        row = dict(rows[i]) if i < len(rows) else {}
        for c, column in enumerate(columns):
            row[str(column.get("key"))] = _projected_value(cells.get((i + 1, c)))
        projected.append(row)
    return projected


def merge_rows_into_sheet(
    columns: list[dict[str, Any]],
    rows: list[dict[str, Any]],
    sheet: list[dict[str, Any]] | None,
) -> list[dict[str, Any]] | None:
    """Agent → sheet: rectangle cells whose rows value differs are overridden
    by the rows value (styling keys preserved); trailing sheet rows beyond
    ``rows`` are dropped inside the rectangle. Returns ``sheet`` unchanged
    when nothing differs. Formula overrides carry the formula only — the
    spreadsheet app recomputes on open, matching the exporter's contract.
    """
    first = sheet[0] if sheet else None
    if not isinstance(first, dict) or not columns:
        return sheet
    cells = cell_map(first)

    overrides: dict[tuple[int, int], dict[str, Any] | None] = {}
    for c, column in enumerate(columns):
        label = column.get("label") or column.get("key")
        if not same_cell_content(label, cells.get((0, c))):
            overrides[(0, c)] = {"v": label, "m": str(label)}
        for i, row in enumerate(rows):
            value = row.get(str(column.get("key")))
            if not same_cell_content(value, cells.get((i + 1, c))):
                overrides[(i + 1, c)] = _value_cell(value)

    def in_rectangle(r: int, c: int) -> bool:
        return c < len(columns) and r >= 1

    ghosts = [key for key in cells if in_rectangle(*key) and key[0] > len(rows)]
    if not overrides and not ghosts:
        return sheet

    merged: list[dict[str, Any]] = []
    for (r, c), existing in cells.items():
        if in_rectangle(r, c) and r > len(rows):
            continue  # agent-deleted row — its formatting goes with it
        if (r, c) not in overrides:
            merged.append({"r": r, "c": c, "v": existing})
            continue
        style = (
            {k: v for k, v in existing.items() if k not in _VALUE_KEYS}
            if isinstance(existing, dict)
            else {}
        )
        next_value = overrides[(r, c)]
        v = style if next_value is None else {**style, **next_value}
        if v:
            merged.append({"r": r, "c": c, "v": v})
    for (r, c), next_value in overrides.items():
        if (r, c) in cells or next_value is None:
            continue
        merged.append({"r": r, "c": c, "v": next_value})
    merged.sort(key=lambda cell: (cell["r"], cell["c"]))

    merged_first = {k: v for k, v in first.items() if k != "data"}
    merged_first["celldata"] = merged
    merged_first["row"] = max(int(first.get("row") or 0), len(rows) + 40)
    merged_first["column"] = max(int(first.get("column") or 0), len(columns) + 2)
    return [merged_first, *list(sheet or [])[1:]]


def _value_cell(rows_value: Any) -> dict[str, Any] | None:
    if rows_value is None or rows_value == "":
        return None
    if _is_formula(rows_value):
        return {"f": rows_value}
    cell: dict[str, Any] = {"v": rows_value, "m": str(rows_value)}
    if isinstance(rows_value, (int, float)) and not isinstance(rows_value, bool):
        cell["ct"] = {"fa": "General", "t": "n"}
    return cell
